### PR TITLE
Add sendmmsg syscall

### PR DESCRIPTION
**Pull request by guzongmin**
sendmmsg is a Linux syscall used by glibc

### CODE REVIEW

File: src/libos/src/net/syscalls.rs [hunk 1]
```diff
@@ -531,6 +531,56 @@ pub fn do_recvmsg(fd: c_int, msg_mut_ptr: *mut msghdr_mut, flags_c: c_int) -> Re
...
+                        msg
+                    } else {
+                        return None;
+                    }
+                };
+
+                socket
+                    .sendmsg(&msg, flags)
+                    .map(|bytes_sent| {
+                        mmsg.msg_len = bytes_sent as u32;
```
Comment (tatetian) [inline src/libos/src/net/syscalls.rs:570]: This should be adding, isn't it?
```rust
mmsg.msg_len += bytes_sent
```

File: src/libos/src/net/syscalls.rs [hunk 2]
```diff
@@ -531,6 +531,56 @@ pub fn do_recvmsg(fd: c_int, msg_mut_ptr: *mut msghdr_mut, flags_c: c_int) -> Re
...
+                        mmsg.msg_len
+                    })
+                    .ok()
+            })
+            .count();
+        Ok(send_count as isize)
+    } else if let Ok(socket) = file_ref.as_unix_socket() {
+        return_errno!(EBADF, "does not support unix socket")
+    } else {
+        return_errno!(EBADF, "not a socket")
```
Comment (tatetian) [inline src/libos/src/net/syscalls.rs:580]: The man page of `sendmsg` states:
>        EBADF  sockfd is not a valid open file descriptor.
>        ENOTSOCK The file descriptor sockfd does not refer to a socket.

So the error code here should be `ENOTSOCK`.

File: src/libos/src/net/syscalls.rs [hunk 3]
```diff
@@ -531,6 +531,56 @@ pub fn do_recvmsg(fd: c_int, msg_mut_ptr: *mut msghdr_mut, flags_c: c_int) -> Re
...
+                    .map(|bytes_sent| {
+                        mmsg.msg_len = bytes_sent as u32;
+                        mmsg.msg_len
+                    })
+                    .ok()
+            })
+            .count();
+        Ok(send_count as isize)
+    } else if let Ok(socket) = file_ref.as_unix_socket() {
+        return_errno!(EBADF, "does not support unix socket")
```
Comment (tatetian) [inline src/libos/src/net/syscalls.rs:578]: The man page of `sendmsg` states:
>        EBADF  sockfd is not a valid open file descriptor
>       EOPNOTSUPP Some bit in the flags argument is inappropriate for the socket type.

We should not use `EBADF` error code when the fd is valid, but the file type does not support this operation. The `EOPNOTSUPP` error code seems to fit our purpose better.

File: src/libos/src/net/syscalls.rs (hunk 1 above)
Comment (guzongmin) [reply src/libos/src/net/syscalls.rs:570]: No. It only indicates the length of single message.

File: src/libos/src/net/syscalls.rs [hunk 4]
```diff
@@ -531,6 +531,56 @@ pub fn do_recvmsg(fd: c_int, msg_mut_ptr: *mut msghdr_mut, flags_c: c_int) -> Re
...
+    from_user::check_ptr(msgvec_ptr)?;
+
+    let mut msgvec = unsafe { std::slice::from_raw_parts_mut(msgvec_ptr, vlen as usize) };
+    let flags = SendFlags::from_bits_truncate(flags_c);
+    let file_ref = current!().file(fd as FileDesc)?;
+
+    if let Ok(socket) = file_ref.as_host_socket() {
+        let send_count = msgvec
+            .iter_mut()
+            .filter_map(|mut mmsg| {
```
[the remark this answers is not in the document]
Comment (guzongmin) [reply src/libos/src/net/syscalls.rs]: Done

File: src/libos/src/net/syscalls.rs (hunk 2 above)
Comment (guzongmin) [reply src/libos/src/net/syscalls.rs]: Done

File: src/libos/src/net/syscalls.rs (hunk 3 above)
Comment (guzongmin) [reply src/libos/src/net/syscalls.rs]: Done

**Review by tatetian — APPROVED** [review]
LGTM

**Pull request merged**